satellite/gracefulexit/endpoint.go

```diff
@@ -0,0 +1,468 @@
+					if err != nil {
+						return Error.Wrap(err)
+					}
+				}
+			}
+		}
+		return nil
+	})
+
+	for {
+		pendingCount := pending.length()
+		// if there are no more transfers and the pending queue is empty, send complete
+		if atomic.LoadInt32(&morePiecesFlag) == 0 && pendingCount == 0 {
+			// TODO check whether failure threshold is met before sending completed
+			// TODO needs exit signature
+			transferMsg := &pb.SatelliteMessage{
+				Message: &pb.SatelliteMessage_ExitCompleted{
+					ExitCompleted: &pb.ExitCompleted{},
+				},
+			}
+			err = stream.Send(transferMsg)
+			if err != nil {
+				return Error.Wrap(err)
+			}
+			break
+		}
+		// skip if there are none pending
+		if pendingCount == 0 {
+			continue
+		}
+
+		request, err := stream.Recv()
+		if err != nil {
+			return eofHandler(err)
+		}
+
+		switch m := request.GetMessage().(type) {
+		case *pb.StorageNodeMessage_Succeeded:
+			err = endpoint.handleSucceeded(ctx, pending, nodeID, m)
+			if err != nil {
+				return Error.Wrap(err)
+			}
+
+		case *pb.StorageNodeMessage_Failed:
+			err = endpoint.handleFailed(ctx, pending, nodeID, m)
+			if err != nil {
+				return Error.Wrap(err)
+			}
+		default:
+			return Error.New("unknown storage node message: %v", m)
+		}
+	}
+
+	if err := group.Wait(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (endpoint *Endpoint) processIncomplete(ctx context.Context, stream processStream, pending *pendingMap, incomplete *TransferQueueItem) error {
+	nodeID := incomplete.NodeID
+	pointer, err := endpoint.metainfo.Get(ctx, string(incomplete.Path))
+	if err != nil {
+		return Error.Wrap(err)
+	}
+	remote := pointer.GetRemote()
+
+	pieces := remote.GetRemotePieces()
+	var nodePiece *pb.RemotePiece
+	excludedNodeIDs := make([]storj.NodeID, len(pieces))
+	for i, piece := range pieces {
+		if piece.NodeId == nodeID && piece.PieceNum == incomplete.PieceNum {
+			nodePiece = piece
+		}
+		excludedNodeIDs[i] = piece.NodeId
+	}
+
+	if nodePiece == nil {
+		endpoint.log.Debug("piece no longer held by node.", zap.String("node ID", nodeID.String()), zap.ByteString("path", incomplete.Path), zap.Int32("piece num", incomplete.PieceNum))
+
+		err = endpoint.db.DeleteTransferQueueItem(ctx, nodeID, incomplete.Path)
+		if err != nil {
+			return Error.Wrap(err)
+		}
+
+		return nil
+	}
+
+	redundancy, err := eestream.NewRedundancyStrategyFromProto(pointer.GetRemote().GetRedundancy())
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	if len(remote.GetRemotePieces()) > redundancy.OptimalThreshold() {
+		endpoint.log.Debug("piece has more pieces than required. removing node from pointer.", zap.String("node ID", nodeID.String()), zap.ByteString("path", incomplete.Path), zap.Int32("piece num", incomplete.PieceNum))
+
+		_, err = endpoint.metainfo.UpdatePieces(ctx, string(incomplete.Path), pointer, nil, []*pb.RemotePiece{nodePiece})
+		if err != nil {
+			return Error.Wrap(err)
+		}
+
+		err = endpoint.db.DeleteTransferQueueItem(ctx, nodeID, incomplete.Path)
```

Comment:
We also need to delete from the piece from the storagenode - and maybe add a test for this case. We can add this TODO to V3-2874